fake_components/src/generic_system.cpp

```diff
@@ -0,0 +1,233 @@
+        {
+          throw std::runtime_error(
+                  "Interface is not found in the standard nor other list. "
+                  "This should never happen!");
+        }
+      }
+    }
+  }
+
+  // Fake sensor command interfaces
+  if (fake_sensor_command_interfaces) {
+    for (uint i = 0; i < info_.sensors.size(); i++) {
+      const auto & sensor = info_.sensors[i];
+      for (const auto & interface : sensor.state_interfaces) {
+        if (!get_interface(
+            sensor.name, sensor_interfaces_, interface.name, i,
+            sensor_fake_commands_, command_interfaces))
+        {
+          throw std::runtime_error(
+                  "Interface is not found in the standard nor other list. "
+                  "This should never happen!");
+        }
+      }
+    }
+  }
+
+  return command_interfaces;
+}
+
+return_type GenericSystem::read()
+{
+  // only do loopback
+  joint_states_ = joint_commands_;
+  other_states_ = other_commands_;
+  if (fake_sensor_command_interfaces) {
+    sensor_states_ = sensor_fake_commands_;
+  }
+  return return_type::OK;
+}
+
+// Private methods
+template<typename HandleType>
+bool GenericSystem::get_interface(
+  const std::string & name,
+  const std::vector<std::string> & interface_list,
+  const std::string & interface_name,
+  const uint vector_index,
+  std::vector<std::vector<double>> & values,
+  std::vector<HandleType> & interfaces)
+{
+  auto it = std::find(interface_list.begin(), interface_list.end(), interface_name);
+  if (it != interface_list.end()) {
+    auto j = std::distance(interface_list.begin(), it);
+    interfaces.emplace_back(name, *it, &values[j][vector_index]);
+    return true;
+  }
+  return false;
+}
+
+void GenericSystem::initialize_storage_vectors(
+  std::vector<std::vector<double>> & commands,
+  std::vector<std::vector<double>> & states,
+  const std::vector<std::string> & interfaces)
+{
+  // Initialize storage for all joints, regardless of their existance
+  commands.resize(interfaces.size());
+  states.resize(interfaces.size());
+  for (uint i = 0; i < interfaces.size(); i++) {
+    commands[i].resize(info_.joints.size(), std::numeric_limits<double>::quiet_NaN());
+    states[i].resize(info_.joints.size(), std::numeric_limits<double>::quiet_NaN());
+  }
+
+  // Initialize with values from URDF
+  for (uint i = 0; i < info_.joints.size(); i++) {
+    const auto & joint = info_.joints[i];
+    for (uint j = 0; j < interfaces.size(); j++) {
+      auto it = joint.parameters.find("initial_" + interfaces[j]);
+      if (it != joint.parameters.end()) {
+        commands[j][i] = std::stod(it->second);
+        states[j][i] = std::stod(it->second);
+      }
+    }
+  }
+}
+
+}  // namespace fake_components
+
+#include "pluginlib/class_list_macros.hpp"
+PLUGINLIB_EXPORT_CLASS(fake_components::GenericSystem, hardware_interface::SystemInterface)
+PLUGINLIB_EXPORT_CLASS(fake_components::GenericRobot, hardware_interface::SystemInterface)
```

Comment:
Out of curiosity, any reason for exporting two different names for the same library .?

Reply:
Only convenience. For the most users in the most cases: System == Robot.

@bmagyar Are you also OK with this?

Reply:
That seems confusing to me. Why exactly?

Reply:
Let's discuss this at the next WG meeting. If we want to revise nomenclature, that's fine, could do that for Galactic but I'd prefer to keep an official nomenclature without convenience duplications.